crates/typst/src/diag.rs

```diff
@@ -228,6 +239,49 @@ impl From<SyntaxError> for SourceDiagnostic {
     }
 }
 
+/// The identifier of a [`SourceDiagnostic`].
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum Identifier {
+    /// Identifier for a built-in compiler warning.
+    Warn(CompilerWarning),
+    /// Identifier for a warning raised by a package.
+    User(EcoString),
+}
+
+/// Built-in compiler warnings.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum CompilerWarning {
```

Comment:
what's the pro/con for using an enum here instead of just using string keys as well?

Reply:
I believe this is the simplest way to have our set of warnings in one place, in a way we can add features to. I had implemented identifier categories using this enum, for example, though I removed it to keep the PR's scope small. We could even generate the warning messages using the enum, but I think this would require some further design discussion (and it could also not be worth it), so I haven't made any efforts in that direction for this PR.
There is also the pro of type-safety when invoking `warn!`, so we can be sure to not accidentally create new identifiers without listing them here first, ensuring that would be an intentional change.

The con is that the identifier and other information are detached from where the warning is actually emitted, but I think it's fine as long as it's in `diag`, or in an appropriate location in general.

Reply:
Switched to just string keys for now after further discussion. We can reconsider enums if we need them in the future.